3rdParty/iresearch/core/utils/memory.cpp

```diff
@@ -75,6 +76,7 @@ Topmost releasable block (keepcost):   %lu\n\
     // output malloc_stats()
     // ...........................................................................
     malloc_stats(); // outputs to stderr
```

Comment:
@gnusi @neunhoef: I would like to get rid of additional compile switches (as we already have too many - we should instead try to remove some switches).
AFAIK the new switch `LIBMUSL_BUILD` is only used to prevent a call to `malloc_stats()`, and for adding library `execinfo`.
I am not sure if we should really call `malloc_stats()` from our executable, as it is not portable anyway. If we want to have information about memory allocations etc. we should probably use external tools for this?

If we could get rid of the dependency to `malloc_stats` that would only leave `execinfo`.
AFAIK we use `execinfo` here in our code:
```
lib/Basics/debugging.cpp:#include <execinfo.h>
3rdParty/iresearch/core/utils/memory.cpp:  #include <execinfo.h> // for backtrace(...)
3rdParty/iresearch/core/utils/log.cpp:  #include <execinfo.h>
```
If we could get rid of execinfo in IResearch, I would also be willing to get rid of the execinfo dependency in ArangoDB core (lib/Basics/debugging.cpp). That would also have the advantage of getting rid of yet another compile switch (ARANGODB_ENABLE_BACKTRACE, triggered by setting `USE_BACKTRACE in CMake).

@gnusi: do you think we can get rid of these two dependencies in IResearch? I will then make sure we get rid of the execinfo dependency in ArangoDB core too.
And that would enable us to get away without any further switch for building a libmusl-based build.

Reply:
@jsteemann I'm not sure we can easily get rid of `malloc_stats()` and `execinfo`. We spent quite a lot of time carefully implementing debugging functionality in IResearch and I confident that most of existing external tools rely on the same tools. @vasiliy-arangodb probably can confirm this.
@neunhoef @jsteemann Am I right that libmusl is the replacement for glibc? If so, what is reason of why do you want to get rid of glibc?

Reply:
What's the problem with `USE_BACKTRACE`, it's pretty useful sometimes

Reply:
build complexity (too many options for building), build dependencies (execinfo)

Reply:
@jsteemann one of potentials options might be to get rid of `execinfo` for release builds

Reply:
which means we would keep if/then/else build complexity?

Reply:
Let me quickly explain about the glibc/libmusl business. With glibc it is essentially not possible to build completely static binaries. However, having completely static binaries would be a *huge* simplification for our release procedures. Imagine we would have only one executable for *all* Linux distributions. This reduces build/release script complexity by two orders of magnitude in scripts size and by two orders of magnitude in build time. This is not an exageration. I am talking about the possibility to build a new release within 15 minutes (at least for Linux) on an arbitrary machine with Docker, rather than the 10h (provided the first try works).

I can now build a completely static executable using alpine linux and libmusl (typically within minutes from scratch if my ccache is loaded), but the problem with this are glibc-only extensions. Currently there are two of them: One is the backtracing functionality, which needs a separate library with libmusl, the other are the malloc-debugging extensions which are simply not there in libmusl (like on OSX and Windows). Currently, I need that additional switch `LIBMUSL_BUILD`, which is ugly and increases complexity.

In general, we are currently drowning in complexity, to the extent that it endangers the existence of this company. Urgent action is needed and I am willing to engage in that action. Some sacrifices will have to be made.

Reply:
@neunhoef @jsteemann 
another way of how we can avoid ugly switches is using SFINAE and/or fake headers carefully maintaining right order of include directories.
E.g. in the example above we can define our own stub for `execinfo.h` and in case if it's not present compiler will use it.

Reply:
I do not like this, it is extremely brittle. The moment the library changes anything we get out of sync with out fake headers and everything is subtly broken. Do we really, really need this functionality? It is not clean POSIX code and thus a liability.

Reply:
@neunhoef 
well, lookup order is well defined so it isn't so fragile as one may imagine at glance.
with SFINAE we can easily make `malloc_stats` optional in iresearch so the question is `execinfo.h` only.
Backtrace functionality (`execinfo.h`) is always platform specific and any tools will eventually use same syscalls under the hood. And yes I would say backtraces are extremely useful and we definitely need them.
In case if you don't like the idea with headers I believe the wisest solution will be to optionally turn it off... at least until we'll come up with the better one

Reply:
I found a macro which is present on glibc but not in libmusl, because it relates to another non-POSIX extension in the malloc area that glibc has but libmusl not. So the usage of mallinfo can be prevented without a separate switch (see latest commits).

Reply:
As to the backtraces: I still do not like us using non-standard libraries. But also for this I found a workaround: cmake has a `find_package(Backtrace)` which we should use anyway, if we use backtrace functionality. I have added this check to the main `CMakeList.txt` (since we are using backtraces anyway). Tests are underway, but linux/glibc, linux/alpine and Windows are fine with this. Currently checking OSX...

Reply:
The current version builds on linux/glibc linux/musl (statically), Windows and OSX. From my perspective we could merge.

Reply:
Still running all unittests one more time...